src/main/java/net/dv8tion/jda/client/entities/impl/ApplicationImpl.java

```diff
@@ -0,0 +1,424 @@
+        {
+            return this.id;
+        }
+
+        @Override
+        public String getInviteUrl(final Collection<Permission> permissions)
+        {
+            return this.getInviteUrl(null, permissions);
+        }
+
+        @Override
+        public String getInviteUrl(final Permission... permissions)
+        {
+            return this.getInviteUrl(null, permissions);
+        }
+
+        @Override
+        public String getInviteUrl(final String guildId, final Collection<Permission> permissions)
+        {
+            StringBuilder builder = new StringBuilder("https://discordapp.com/oauth2/authorize?client_id=");
+            builder.append(this.getId());
+            builder.append("&scope=bot");
+            if (permissions != null && !permissions.isEmpty())
+            {
+                builder.append("&permissions=");
+                builder.append(Permission.getRaw(permissions));
+            }
+            if (guildId != null)
+            {
+                builder.append("&guild_id=");
+                builder.append(guildId);
+            }
+            return builder.toString();
+        }
+
+        @Override
+        public String getInviteUrl(final String guildId, final Permission... permissions)
+        {
+            return this.getInviteUrl(guildId, permissions == null ? null : Arrays.asList(permissions));
+        }
+
+        public JDA getJDA()
+        {
+            return ApplicationImpl.this.getJDA();
+        }
+
+        @Override
+        public String getName()
+        {
+            return this.name;
+        }
+
+        @Override
+        public String getToken()
+        {
+            return this.token;
+        }
+
+        @Override
+        public int hashCode()
+        {
+            return this.id.hashCode();
+        }
+
+        @Override
+        public RestAction<Bot> resetToken()
+        {
+            Route.CompiledRoute route = Route.Applications.RESET_BOT_TOKEN.compile(this.id);
+            return new RestAction<Bot>(getJDA(), route, null)
+            {
+                @Override
+                protected void handleResponse(final Response response, final Request request)
+                {
+                    if (response.isOk())
+                        request.onSuccess(BotImpl.this.updateFromJson(response.getObject()));
+                    else
+                        request.onFailure(response);
+                }
+            };
+        }
+
+        @Override
+        public String toString()
+        {
+            return "Application.Bot(" + this.id + ")";
+        }
+
+        public BotImpl updateFromJson(final JSONObject object)
+        {
+            this.name = object.getString("username");
+            this.discriminator = object.getString("discriminator");
+            this.token = object.getString("token");
+            this.id = object.getString("id");
+            this.avatarId = object.has("avatar") ? null : object.getString("avatar");
```

Comment:
The ternary needs to be flipped.
In addition: Shouldn't all fields here check `JSONObject#has` since this is called  `update` not `create` it might receive only partial data